File: ios/Sources/LaunchNavigatorPlugin/LaunchNavigatorPlugin.swift
import Foundation
import Capacitor
import UIKit
import MapKit

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(LaunchNavigatorPlugin)
public class LaunchNavigatorPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "8.0.23"
    public let identifier = "LaunchNavigatorPlugin"
    public let jsName = "LaunchNavigator"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "navigate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isAppAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAvailableApps", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSupportedApps", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDefaultApp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAppIcons", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "refreshAppIcons", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearIconCache", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = LaunchNavigator()
    private let iconCacheQueue = DispatchQueue(label: "app.capgo.plugin.launch_navigator.icons")

    override public func load() {
        super.load()
        implementation.webPathResolver = { [weak self] localURL in
            guard let portable = self?.bridge?.portablePath(fromLocalURL: localURL) else {
                return nil
            }
            return portable.absoluteString
        }
    }

    @objc func navigate(_ call: CAPPluginCall) {
        guard let destination = call.getArray("destination") as? [Double],
              destination.count == 2 else {
            call.reject("Destination coordinates [latitude, longitude] are required")
            return
        }

        let lat = destination[0]
        let lon = destination[1]
        let options = call.getObject("options") ?? [:]

        let app = options["app"] as? String ?? "apple_maps"
        let transportMode = options["transportMode"] as? String ?? "driving"

        var start: CLLocationCoordinate2D?
        if let startCoords = options["start"] as? [Double], startCoords.count == 2 {
            start = CLLocationCoordinate2D(latitude: startCoords[0], longitude: startCoords[1])
        }

        let startName = options["startName"] as? String
        let destinationName = options["destinationName"] as? String

        DispatchQueue.main.async {
            self.implementation.navigate(
                app: app,
                destination: CLLocationCoordinate2D(latitude: lat, longitude: lon),
                start: start,
                startName: startName,
                destinationName: destinationName,
                transportMode: transportMode,
                viewController: self.bridge?.viewController
            ) { success, error in
                if success {
                    call.resolve()
                } else {
                    call.reject(error ?? "Failed to launch navigation app")
                }
            }
        }
    }

    @objc func isAppAvailable(_ call: CAPPluginCall) {
        guard let app = call.getString("app") else {
            call.reject("App identifier is required")
            return
        }

        let available = implementation.isAppAvailable(app: app)
        call.resolve([
            "available": available
        ])
    }

    @objc func getAvailableApps(_ call: CAPPluginCall) {
        let apps = implementation.getAvailableApps()
        call.resolve([
            "apps": apps
        ])
    }

    @objc func getSupportedApps(_ call: CAPPluginCall) {
        let apps = implementation.getSupportedApps()
        call.resolve([
            "apps": apps
        ])
    }

    @objc func getDefaultApp(_ call: CAPPluginCall) {
        call.resolve([
            "app": "apple_maps"
        ])
    }

    @objc func getAppIcons(_ call: CAPPluginCall) {
        let options = call.options as? [String: Any] ?? [:]
        iconCacheQueue.async {
            let result = self.implementation.getAppIcons(options: options, forceRefresh: false)
            DispatchQueue.main.async {
                call.resolve(result)
            }
        }
    }

    @objc func refreshAppIcons(_ call: CAPPluginCall) {
        let options = call.options as? [String: Any] ?? [:]
        iconCacheQueue.async {
            let result = self.implementation.getAppIcons(options: options, forceRefresh: true)
            DispatchQueue.main.async {
                call.resolve(result)
            }
        }
    }

    @objc func clearIconCache(_ call: CAPPluginCall) {
        let options = call.options as? [String: Any] ?? [:]
        iconCacheQueue.async {
            let result = self.implementation.clearIconCache(options: options)
            DispatchQueue.main.async {
                call.resolve(result)
            }
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }

}

File: android/src/main/java/app/capgo/plugin/launch_navigator/LaunchNavigatorPlugin.java
package app.capgo.plugin.launch_navigator;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "LaunchNavigator")
public class LaunchNavigatorPlugin extends Plugin {

    private final String pluginVersion = "7.1.9";

    private LaunchNavigator implementation;

    @Override
    public void load() {
        implementation = new LaunchNavigator(getContext());
    }

    @PluginMethod
    public void navigate(PluginCall call) {
        JSArray destination = call.getArray("destination");
        if (destination == null || destination.length() != 2) {
            call.reject("Destination coordinates [latitude, longitude] are required");
            return;
        }

        try {
            double lat = destination.getDouble(0);
            double lon = destination.getDouble(1);

            JSObject options = call.getObject("options", new JSObject());
            String app = options.getString("app", "google_maps");
            String transportMode = options.getString("transportMode", "driving");

            Double startLat = null;
            Double startLon = null;

            // Try to get start array as JSONArray from the options object
            try {
                org.json.JSONArray startArray = options.getJSONArray("start");
                if (startArray != null && startArray.length() == 2) {
                    startLat = startArray.getDouble(0);
                    startLon = startArray.getDouble(1);
                }
            } catch (Exception e) {
                // start is optional, ignore if not present or invalid
            }

            String startName = options.getString("startName");
            String destinationName = options.getString("destinationName");

            boolean result = implementation.navigate(app, lat, lon, startLat, startLon, startName, destinationName, transportMode);

            if (result) {
                call.resolve();
            } else {
                call.reject("Failed to launch navigation app");
            }
        } catch (Exception e) {
            call.reject("Error parsing navigation parameters: " + e.getMessage());
        }
    }

    @PluginMethod
    public void isAppAvailable(PluginCall call) {
        String app = call.getString("app");
        if (app == null) {
            call.reject("App identifier is required");
            return;
        }

        boolean available = implementation.isAppAvailable(app);

        JSObject ret = new JSObject();
        ret.put("available", available);
        call.resolve(ret);
    }

    @PluginMethod
    public void getAvailableApps(PluginCall call) {
        JSArray apps = implementation.getAvailableApps();

        JSObject ret = new JSObject();
        ret.put("apps", apps);
        call.resolve(ret);
    }

    @PluginMethod
    public void getSupportedApps(PluginCall call) {
        JSArray apps = implementation.getSupportedApps();

        JSObject ret = new JSObject();
        ret.put("apps", apps);
        call.resolve(ret);
    }

    @PluginMethod
    public void getDefaultApp(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("app", "google_maps");
        call.resolve(ret);
    }

    @PluginMethod
    public void getPluginVersion(final PluginCall call) {
        try {
            final JSObject ret = new JSObject();
            ret.put("version", this.pluginVersion);
            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not get plugin version", e);
        }
    }
}
